### DISCUSSION TITLE
I think I still don't understand dependencies

### DISCUSSION THREAD

**Discussion by davidchisnall** [xmake-io/xmake#2799] [Q&A]
I have a target that does:

```
add_deps(...)
```

This target has a rule that implements `on_linkcmd` to define the link job.  This batchcmd is sometimes (deterministically, 100% of the time, on first build after clean) run before some of the dependencies listed in `add_deps` are built, which then causes the linker to fail because some files don't exist.  

I am using `batchcmd:add_depfiles` in my rule to specify some additional files, but I'd assumed that the dependencies are also built.  I've also tried explicitly using `add_depfiles` on the set of files that are passed to the linker.  The files being passed to the linker are collected with:

```
for name, dep in pairs(target:deps())
  -- filter out some and use dep:targetfile() to get the files for the rest
end
```

I don't understand why a target can be built before its dependencies are built.

**Comment by waruqi** [comment]
I'm not sure what you are talking about, you could provide a reproducible demo project and then provide some error output.

**Reply by davidchisnall** [reply]
I can't yet open source this build system because the project is not open source and I'm not sure what a minimal reproducer will look like.  That's why I'm asking how dependencies are supposed to work: how does the system track dependencies between targets?  What might I have got wrong?

I have been exploring using xmake for [this project](http://aka.ms/smallestcheri).  We're currently using CMake and I hoped xmake would be better.

**Reply by waruqi** [reply]
Without any xmake.lua configuration and error messages, I can't provide much more help. I need to get some xmake configuration snippets and explain where it doesn't work as expected so I can see how you're using it.

**Reply by davidchisnall** [reply]
What do you need beyond what I provided?  

 - Is it expected that, if a target is defined to depend on another via `add_deps` in declaration scope that the target won't be built until the dependency is built (this is my expectation, is it correct)?
 - Are `on_linkcmds` functions in rules meant to explicitly provide some form of barrier to the `batchcmds` to prevent them from running until the dependencies are built?
 - Is `add_depfiles` meant to provide this barrier, or is there some other thing?
 - Is there a mechanism for dumping what xmake thinks the dependency graph looks like so that I can see where things are missing?

**Reply by waruqi** [reply]
> Is it expected that, if a target is defined to depend on another via add_deps in declaration scope that the target won't be built until the dependency is built (this is my expectation, is it correct)?

Not exactly, it is possible that the source files of the dependent target and the current target will be compiled at the same time, as there are usually no dependencies between the source files, which can speed up compilation.

Unless you have set the `set_policy("build.across_targets_in_parallel", false)`, https://xmake.io/#/manual/project_target?id=targetset_policy

But the linking process will be executed strictly in dependency order.

> Are on_linkcmds functions in rules meant to explicitly provide some form of barrier to the batchcmds to prevent them from running until the dependencies are built?

It is always executed after the dependency target build link is completed.

> Is add_depfiles meant to provide this barrier, or is there some other thing?

No, It is usually used for incremental compilation, or to avoid repeated secondary compilation, and will only re-execute batchcmds when the dependency files are updated

> Is there a mechanism for dumping what xmake thinks the dependency graph looks like so that I can see where things are missing?

no way now.

**Reply by davidchisnall** [reply]
> Not exactly, it is possible that the source files of the dependent target and the current target will be compiled at the same time, as there are usually no dependencies between the source files, which can speed up compilation.

That's fine.  It's the output that's the problem.

> Unless you have set the set_policy("build.across_targets_in_parallel", false), https://xmake.io/#/manual/project_target?id=targetset_policy

Nope, I haven't touched that.

> But the linking process will be executed strictly in dependency order.

That's what I'd hoped, so there must be something wrong in my setup.

> It is always executed after the dependency target build link is completed.

The dependency also has a custom link step.  Is there a way in which that might be a problem (do I need to explicitly report that the target has finished in some way)?

These are the (WIP) rules that are used for the dependencies:

```lua
rule("cherimcu.component")
        on_load(function (target)
                -- Static library
                target:set("kind", "static")
                -- Suffix, for now, is a compartment (eventually, libraries should have
                -- a different suffix)
                target:set("extension", ".compartment")
                -- No prefix
                target:set("prefixname", "")
                -- Default to C2x for C code.  This works around a bug where clatest
                -- doesn't do the right thing on the latest xmake release.
                target:add("cflags", "-std=c2x")
                -- Mark this target as a CHERI MCU component.
        end)
        on_linkcmd(function (target, batchcmds, opt)
                -- Get a specified linker script, or set the default to the compartment
                -- linker script.
                local linkerscript_name = target:get("ldscript") or "compartment.ldscript"
                local linkerscript = path.join(scriptdir, linkerscript_name)
                -- Link using the compartment's linker script.
                batchcmds:add_depfiles(linkerscript)
                batchcmds:show_progress(opt.progress, "linking compartment " .. target:filename())
                batchcmds:mkdir(target:targetdir())
                batchcmds:vrunv(target:tool("ld"), table.join({"--script=" .. linkerscript, "--compartment", "--no-relax", "-o", target:targetfile()}, target:objectfiles()), opt)
        end)

-- CHERI MCU libraries are currently built as compartments, without a
-- `-cheri-compartment` flag.
rule("cherimcu.library")
        add_deps("cherimcu.component")
        on_load(function (target)
                -- Mark this target as a CHERI MCU library.
                target:set("cherimcu.type", "library")
        end)

-- CHERI MCU compartments have an explicit compartment name passed tot he
-- compiler.
rule("cherimcu.compartment")
        add_deps("cherimcu.component")
        on_load(function (target)
                -- Mark this target as a CHERI MCU compartment.
                target:set("cherimcu.type", "compartment")
        end)
        after_load(function (target)
                local compartment = target:get("cherimcu.compartment") or target:name()
                target:add("cxflags", "-cheri-compartment=" .. compartment)
        end)
```

One of the ones that isn't being built before the final target looks just like this:

```lua
target("lib")
        add_rules("cherimcu.library")
        add_files({a load of C files})
```

The final firmware build adds these as dependencies using `add_deps` in description scope.  It sets its type as `binary` and then provides a custom `on_linkcmds` that looks like this:

```lua
        on_linkcmd(function (target, batchcmds, opt)
                import("core.project.config")
                -- Get a specified linker script, or set the default to the compartment
                -- linker script.
                local linkerscript = path.join(config.buildir(), target:name() .. "-firmware.ldscript")
                -- Link using the firmware's linker script.
                batchcmds:add_depfiles(linkerscript)
                batchcmds:show_progress(opt.progress, "linking firmware " .. target:filename())
                batchcmds:mkdir(target:targetdir())
                local objects = target:objectfiles()
                for name, dep in pairs(target:deps()) do
                        if (dep:get("cherimcu.type") == "library") or
                                (dep:get("cherimcu.type") == "compartment") or
                                (dep:get("cherimcu.type") == "privileged-compartment") then
                                table.insert(objects, dep:targetfile())
                        end
                end
                batchcmds:add_depfiles(objects)
                batchcmds:vrunv(target:tool("ld"), table.join({"--script=" .. linkerscript, "--no-relax", "-o", target:targetfile()}, objects), opt)
                batchcmds:show_progress(opt.progress, "Creating firmware dump" .. target:filename())
                batchcmds:vrunv(target:tool("objdump"), {"-glxsdrS", target:targetfile()}, table.join(opt, {stdout = target:targetfile() .. ".dump"}))
        end)
```

**Reply by waruqi** [reply]
You are confusing `rule.add_deps` and `target.add_deps`, which are different.

`rule.add_deps` is used to control the dependencies between rules, if `rule(a).add_deps(b)` is configured then `add_rules(a)` will apply both rule a and b to target.

Whereas `target.add_deps` is what controls the build dependencies between build targets, if you configure it in rule you should use `target:add("deps", "xxx")`

I don't see that you have configured any target.add_deps.

**Reply by davidchisnall** [reply]
Sorry, when the firmware target is created it uses (description scope) `add_deps` to add the dependencies.  This is why they then show up in the `target:deps()` iteration in the rule target.

**Reply by waruqi** [reply]
But right now I'm not sure what target dependencies you've configured, or what errors you've encountered, and I don't have any error logs that I can analyse or an example project that I can debug.

If you think there's a problem with the build order, you'll still need to provide a minimal example project that can reproduce the problem so I can continue to analyse the cause.

**Reply by davidchisnall** [reply]
That may have to wait until our open source release, which unfortunately means that we'll have to keep using the CMake build system and reconsider xmake in a year or so.  I can run whatever debugging commands you want but I can't share the code yet.  Thanks anyway.

**Reply by waruqi** [reply]
I don't need your complete code, huge code and environments are hard to debug. Provide minimal reproducible example projects that just contain empty code and configuration scripts that reproduce build order errors, and then tell me what errors I encounter.

**Reply by davidchisnall** [reply]
I can't produce a reduced test case without some idea of what kinds of thing might be causing this.  It seems to have gone away after some changes that I'd expect to be completely unrelated.  It also seems not to happen at all when I run `xmake` from the source directory.
